crates/era-cheatcodes/src/cheatcodes.rs

```diff
@@ -1357,8 +1479,114 @@ impl CheatcodeTracer {
+        reason: Option<&Vec<u8>>,
+        op: zkevm_opcode_defs::RetOpcode,
+        state: &VmLocalStateData<'_>,
+        memory: &SimpleMemory<H>,
+    ) -> Result<(), Vec<u8>> {
+        match (op, reason) {
+            (zkevm_opcode_defs::RetOpcode::Revert, Some(expected_reason)) => {
+                let retdata = {
+                    let ptr = state.vm_local_state.registers
+                        [RET_IMPLICIT_RETURNDATA_PARAMS_REGISTER as usize];
+                    assert!(ptr.is_pointer);
+                    let fat_data_pointer = FatPointer::from_u256(ptr.value);
+                    memory.read_unaligned_bytes(
+                        fat_data_pointer.memory_page as usize,
+                        fat_data_pointer.start as usize,
+                        fat_data_pointer.length as usize,
+                    )
+                };
+
+                tracing::debug!(?expected_reason, ?retdata);
+                if !expected_reason.is_empty() && retdata.is_empty() {
+                    return Err("call reverted as expected, but without data".to_string().into())
+                }
+
+                match VmRevertReason::from(retdata.as_slice()) {
+                    VmRevertReason::General { msg, data: _ } => {
+                        let expected_reason = String::from_utf8_lossy(expected_reason).to_string();
+                        if msg == expected_reason {
+                            Ok(())
+                        } else {
+                            Err(format!(
+                                "Error != expected error: {} != {}",
+                                &msg, expected_reason,
+                            )
+                            .into())
+                        }
+                    }
+                    VmRevertReason::Unknown { function_selector: _, data } => {
+                        if &data == expected_reason {
+                            Ok(())
+                        } else {
+                            Err(format!(
+                                "Error != expected error: {:?} != {:?}",
+                                &data, expected_reason,
+                            )
+                            .into())
+                        }
+                    }
+                    _ => {
+                        tracing::error!("unexpected revert reason");
+                        Err("unexpected revert reason".to_string().into())
+                    }
+                }
+            }
+            (zkevm_opcode_defs::RetOpcode::Revert, None) => {
+                tracing::debug!("any revert accepted");
+                Ok(())
+            }
+            (zkevm_opcode_defs::RetOpcode::Ok, _) => {
+                tracing::debug!("expected revert but call succeeded");
+                Err("expected revert but call succeeded".to_string().into())
+            }
+            (zkevm_opcode_defs::RetOpcode::Panic, _) => todo!("ignore/return error ?"),
```

Comment:
Since we are having a noop on `handle_return` for `RetOpcode::Panic`, we want it to not affect the execution flow at all, as such this function must not be called in a panic - so we can panic here as well to make this future proof that it's not called during panic :)